docs/csharp/programming-guide/concepts/expression-trees/how-to-use-expression-trees-to-build-dynamic-queries.md

```diff
@@ -1,107 +1,266 @@
+
+## IQueryable, IQueryable\<T> and expression trees
+
+Fundamentally, an <xref:System.Linq.IQueryable> has two components:
+
+* <xref:System.Linq.IQueryable.Expression>&mdash;a language- and datasource-agnostic representation of the current query's elements, in the form of an expression tree.
+* <xref:System.Linq.IQueryable.Provider>&mdash;an object which knows how to translate the current query into actual .NET objects, a.k.a. an instance of a LINQ provider.
+
+In the context of dynamic querying, the provider will usually remain the same; the expression tree of the query will be different from query to query.
+
+Because expression trees are immutable, if you want a different expression tree&mdash;and thus a different query&mdash;you'll need to translate the existing expression tree to a new one, and thus to a new **IQueryable**.
+
+The following sections describe specific techniques for querying differently in response to runtime state:
+
+- Referencing runtime state directly in the expression tree
+- Calling additional LINQ methods
+- Varying the expression tree passed into the LINQ methods
+- Constructing expression trees using the factory methods at <xref:System.Linq.Expressions.Expression>
+
+## Referencing runtime state from the expression tree
+
+Assuming the LINQ provider supports it, the simplest way to query dynamically is to reference the runtime state directly in the query, via a closed-over variable:
+
+```csharp
+var length = 1;
+var qry = companyNamesSource
+    .Select(x => x.Substring(0, length))
+    .Distinct();
+    
+Console.WriteLine(string.Join(",", qry));
+// prints: C, A, S, W, G, H, M, N, B, T, L, F
+
+length = 2;
+Console.WriteLine(string.Join(",", qry));
+// prints: Co, Al, So, Ci, Wi, Gr, Ad, Hu, Wo, Ma, No, Bl, Tr, Th, Lu, Fo 
+```
+
+Note that the expression tree hasn't been modified; the query returns different values only because the value of `length` has been changed.
+
+## Calling additional LINQ methods
+
+The LINQ methods on <xref:System.Linq.Queryable> which don't execute the query but rather return a new translated query object, generally consist of two steps:
+
+* wrap the current expression tree in a <xref:System.Linq.Expressions.MethodCallExpression> representing the method call, and
+* pass the newly wrapped expression tree back into the provider for optional additional processing.
+
+For example, the source code for <xref:System.Linq.Queryable.Take> looks something like this:
+
+```csharp
+public static IQueryable<TSource> Take<TSource>(this IQueryable<TSource> source, int count) {
+
+    // wrap the original expression in a MethodCallExpression using the Call factory method    
+    var wrapped =
+        Expression.Call(
+            null,
+            CachedReflectionInfo.Take_TSource_2(typeof(TSource)),
```

Comment:
This type and method is not defined in this article. You also don't tie its use to the later example that composes a new query. We need to explain more about that process in this section.

Reply:
Agreed.

I'm trying to emphasize the idea that the `IQueryable<T>`-returnning methods do nothing more than translate one query object into another, by showing the [source](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs#L380) for `Queryable.Take`. The problem is, it's very hard to see this behind the exception handling and caching code.

I'll edit this out in a later commit.

Reply:
Fixed in cc6550ea6ce39020e8838d4ff08fb2fb59232a73